internal/controller/ec2.aws/securitygroup_controller.go

```diff
@@ -289,60 +322,18 @@ func (r *SecurityGroupReconciliation) reconcileNormal(ctx context.Context) (ctrl
 		}
 	}
 
-	for _, infrastructureRef := range r.sg.Spec.InfrastructureRef {
-		switch infrastructureRef.Kind {
-		case "KopsMachinePool":
-			kmp := kinfrastructurev1alpha1.KopsMachinePool{}
-			key := client.ObjectKey{
-				Name:      infrastructureRef.Name,
-				Namespace: infrastructureRef.Namespace,
-			}
-			if err := r.Client.Get(ctx, key, &kmp); err != nil {
-				return resultError, err
-			}
-
-			kcp := &kcontrolplanev1alpha1.KopsControlPlane{}
-			key = client.ObjectKey{
-				Name:      kmp.Spec.ClusterName,
-				Namespace: kmp.ObjectMeta.Namespace,
-			}
-			if err := r.Client.Get(ctx, key, kcp); err != nil {
-				return resultError, err
-			}
-			err := r.attachKopsMachinePool(ctx, csg, kcp, kmp)
-			if err != nil {
-				if errors.Is(err, ErrSecurityGroupNotAvailable) {
-					return requeue30seconds, nil
-				}
-				return resultError, err
-			}
-		case "KopsControlPlane":
-			kcp := &kcontrolplanev1alpha1.KopsControlPlane{}
-			key := client.ObjectKey{
-				Name:      infrastructureRef.Name,
-				Namespace: infrastructureRef.Namespace,
-			}
-			if err := r.Client.Get(ctx, key, kcp); err != nil {
-				return resultError, err
-			}
+	result, err := r.ensureAttachReferences(ctx, csg)
+	if err != nil {
+		return result, err
+	}
 
-			kmps, err := kops.GetKopsMachinePoolsWithLabel(ctx, r.Client, "cluster.x-k8s.io/cluster-name", kcp.Name)
-			if err != nil {
-				return resultError, err
-			}
-			err = r.attachKopsMachinePool(ctx, csg, kcp, kmps...)
-			if err != nil {
-				if errors.Is(err, ErrSecurityGroupNotAvailable) {
-					return requeue30seconds, nil
-				}
-				return resultError, err
-			}
-		default:
-			return resultError, fmt.Errorf("infrastructureRef not supported")
-		}
+	result, err = r.ensureDetachRemovedReferences(ctx, csg)
+	if err != nil {
+		return result, err
 	}
 
 	r.sg.Status.Ready = true
+	r.sg.Status.AppliedInfrastructureRef = r.sg.DeepCopy().Spec.InfrastructureRef
```

Comment:
why do we need to call this DeepCopy method here

Reply:
No specific reason, I guess we saw that in another place and just copied without properly judging the why

Reply:
But we can just assign the value directly